### DISCUSSION TITLE
BoostAsioScheduler not compatible with rocketmq-client

### DISCUSSION THREAD

**Discussion by qazwsxedckll** [daniele77/cli#140] [General]
When using BoostAsioScheduler, rocketmq-client(https://github.com/apache/rocketmq-client-cpp) will have segment fault after start in a few seconds. rocketmq_client uses boost_1_58_0 and asio.

I have read your instructions and boost::asio::io_context ioc is not access to me, 

#0  0x000000000072f272 in boost::asio::detail::op_queue_access::next<boost::asio::detail::reactor_op> (
    o=0x7265706f72702e73) at /usr/local/include/boost/asio/detail/op_queue.hpp:35
#1  0x00000000007282e4 in boost::asio::detail::op_queue<boost::asio::detail::reactor_op>::pop (this=0x7fe8700059e0)
    at /usr/local/include/boost/asio/detail/op_queue.hpp:97
#2  0x00000000007283e6 in boost::asio::detail::op_queue<boost::asio::detail::reactor_op>::~op_queue (
    this=0x7fe8700059e0, __in_chrg=<optimized out>) at /usr/local/include/boost/asio/detail/op_queue.hpp:80
#3  0x0000000000734b39 in boost::asio::detail::epoll_reactor::descriptor_state::~descriptor_state (
    this=0x7fe870005960, __in_chrg=<optimized out>) at /usr/local/include/boost/asio/detail/epoll_reactor.hpp:57
#4  0x0000000000734b7d in boost::asio::detail::object_pool_access::destroy<boost::asio::detail::epoll_reactor::descriptor_state> (o=0x7fe870005960) at /usr/local/include/boost/asio/detail/object_pool.hpp:47
#5  0x000000000072f1c8 in boost::asio::detail::object_pool<boost::asio::detail::epoll_reactor::descriptor_state>::destroy_list (this=0x7fe884000c60, list=0x76726553656d614e) at /usr/local/include/boost/asio/detail/object_pool.hpp:156
#6  0x0000000000727fcc in boost::asio::detail::object_pool<boost::asio::detail::epoll_reactor::descriptor_state>::~object_pool (this=0x7fe884000c60, __in_chrg=<optimized out>) at /usr/local/include/boost/asio/detail/object_pool.hpp:78
#7  0x000000000071d73e in boost::asio::detail::epoll_reactor::~epoll_reactor (this=0x7fe884000ba0,
    __in_chrg=<optimized out>) at /usr/local/include/boost/asio/detail/impl/epoll_reactor.ipp:66
#8  0x000000000071d798 in boost::asio::detail::epoll_reactor::~epoll_reactor (this=0x7fe884000ba0,
    __in_chrg=<optimized out>) at /usr/local/include/boost/asio/detail/impl/epoll_reactor.ipp:72
#9  0x00000000009643ce in rocketmq::TopicPublishInfo::~TopicPublishInfo() ()
#10 0x00000000009645f7 in boost::detail::sp_counted_impl_p<rocketmq::TopicPublishInfo>::dispose() ()
#11 0x000000000094ecca in boost::detail::sp_counted_base::release() [clone .part.0] ()
#12 0x00000000009654cd in std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, boost::shared_ptr<rocketmq::TopicPublishInfo> >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, boost::shared_ptr<rocketmq::TopicPublishInfo> > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, boost::shared_ptr<rocketmq::TopicPublishInfo> > > >::erase(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#13 0x00000000009538d6 in rocketmq::MQClientFactory::addTopicInfoToTable(std::__cxx11::basic_string<char, std::char_trai---Type <return> to continue, or q <return> to quit---
ts<char>, std::allocator<char> > const&, boost::shared_ptr<rocketmq::TopicPublishInfo>) ()
#14 0x0000000000956166 in rocketmq::MQClientFactory::updateTopicRouteInfoFromNameServer(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rocketmq::SessionCredentials const&, bool) ()
#15 0x00000000009591f1 in rocketmq::MQClientFactory::updateTopicRouteInfo(boost::system::error_code&, boost::shared_ptr<boost::asio::basic_deadline_timer<boost::posix_time::ptime, boost::asio::time_traits<boost::posix_time::ptime>, boost::asio::deadline_timer_service<boost::posix_time::ptime, boost::asio::time_traits<boost::posix_time::ptime> > > >) ()
#16 0x000000000096913f in boost::asio::detail::wait_handler<boost::_bi::bind_t<void, boost::_mfi::mf2<void, rocketmq::MQClientFactory, boost::system::error_code&, boost::shared_ptr<boost::asio::basic_deadline_timer<boost::posix_time::ptime, boost::asio::time_traits<boost::posix_time::ptime>, boost::asio::deadline_timer_service<boost::posix_time::ptime, boost::asio::time_traits<boost::posix_time::ptime> > > > >, boost::_bi::list3<boost::_bi::value<rocketmq::MQClientFactory*>, boost::_bi::value<boost::system::error_code>, boost::_bi::value<boost::shared_ptr<boost::asio::basic_deadline_timer<boost::posix_time::ptime, boost::asio::time_traits<boost::posix_time::ptime>, boost::asio::deadline_timer_service<boost::posix_time::ptime, boost::asio::time_traits<boost::posix_time::ptime> > > > > > > >::do_complete(boost::asio::detail::task_io_service*, boost::asio::detail::task_io_service_operation*, boost::system::error_code const&, unsigned long) ()
#17 0x000000000095edfc in boost::asio::detail::task_io_service::run(boost::system::error_code&) ()
#18 0x0000000000958b09 in rocketmq::MQClientFactory::startScheduledTask(bool) ()
#19 0x0000000000991bf3 in thread_proxy ()
#20 0x00007fe89af20fa3 in ?? ()
#21 0x0000000000000000 in ?? ()

**Comment by daniele77** [comment]
Not knowing your code, it's pretty hard to give you any advice.
How do you use `cli`? Are you using the `boost asio io_service` instantiated by `rocketmq-client`? Or are you instantiate a second `io_service`? Or a `LoopScheduler`?

Some hints:
- I'm not sure `cli` library can work with `boost `version 1.58. I tested it with `boost `version >= 1.60
- cli can work with `asio `versions having io_service instead of io_context: cli detects asio version and use the correct one
- the best solution would be to pass to cli the `io_service `of the `rocketmq-client` library (I guess `rocketmq-client` has a way to export its io_service instance reference). If the `rocketmq-client` library uses a boost version too old for cli, you can try to compile `rocketmq-client` with `boost `>= 1.60
- if the previous solution doesn't work, you can try to use cli `LoopScheduler `for the cli, if you don't need remote shell, or standalone asio just for the `cli` library

BTW: If you're willing to github sponsorship this project, I can give you a consultation on your code.

**Reply by qazwsxedckll** [reply]
- I am using cli the same way in examples/complete.cpp with BoostAsioScheduler.
- It is a little trickier, rocketmq-client will download it's own boost when building and it is not the same using by the rest. It will instantiate its own io_service I think.
- I am using https://github.com/apache/rocketmq-client-cpp/blob/master/example/PushConsumer.cpp this example without the producer.
- I haven't found a way to get the io_service instance.
- LoopScheduler works fine.
- I am glad to be your first sponsor~
- This is not an urgent problem. I will try to solve by myself first. Thanks for your reply.

**Reply by daniele77** [reply]
I'm glad you are the first sponsor of this library. I think it's an useful project, and many developers are using it, but no sponsor until now :-/

Anyway. If you want to share in some way the code producing segmentation fault (a minimal project, please) I will try to help you.

My first guess is a concurrency issue. If there are two instances of `io_service` in the project (one inside `rocketmq-client` and the other instantiated by you for `cli`), their `run()` methods are surely called by two different threads. This means that you should provide some kind of concurrency protection mechanism at the boundary of the two threads.
Please note that the same applies if you have one `io_service` (inside `rocketmq-client` library) and one `LoopScheduler` (for cli).
